src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,144 @@
 
-    @Override
-    public void put(K key, V value) {
+    private Node<K, V>[] table;
+    private int size;
+    private int threshold;
+    private final float loadFactor;
+
+    public MyHashMap() {
+        this(DEFAULT_INITIAL_CAPACITY, DEFAULT_LOAD_FACTOR);
+    }
 
+    public MyHashMap(int initialCapacity, float loadFactor) {
+        if (initialCapacity <= 0) {
+            throw new IllegalArgumentException("Initial capacity must be positive");
+        }
+        if (loadFactor <= 0 || Float.isNaN(loadFactor)) {
+            throw new IllegalArgumentException("Load factor must be positive");
+        }
+        this.loadFactor = loadFactor;
+        threshold = (int) (initialCapacity * loadFactor);
+        table = new Node[initialCapacity];
     }
 
     @Override
     public V getValue(K key) {
+        int hash = hash(key);
+        int index = indexFor(hash, table.length);
+        Node<K, V> node = table[index];
+        while (node != null) {
+            if (node.key == null || node.key.equals(key)) {
+                return node.value;
+            }
+            node = node.next;
+        }
         return null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    @Override
+    public void put(K key, V value) {
+        if (size >= threshold) {
+            resize();
+        }
+        if (key == null) {
+            putForNullKey(value);
+        } else {
+            putNonNullKey(key, value);
+        }
+    }
+
+    private int hash(Object key) {
+        return key == null ? 0 : key.hashCode();
+    }
+
+    private void putForNullKey(V value) {
+        if (table[0] != null) {
+            Node<K, V> element = table[0];
+            if (element.key != null && element.next == null) {
+                element.next = new Node<>(0, null, value, null);
+                size++;
+            } else {
+                while (element != null) {
+                    if (element.key == null) {
+                        element.value = value;
+                    }
+                    element = element.next;
+                }
+            }
+        } else {
+            table[0] = newNode(0, null, value, null);
+            size++;
+        }
+    }
+
+    private void putNonNullKey(K key, V value) {
+        int hash = hash(key);
+        int index = indexFor(hash, table.length);
+        for (Node<K, V> element = table[index]; element != null; element = element.next) {
+            Object k = element.key;
+            if (element.hash == hash && (k == key || key.equals(k))) {
+                element.value = value;
+                return;
+            }
+        }
+        createEntry(hash, key, value, index);
+    }
+
+    private void createEntry(int hash, K key, V value, int bucketIndex) {
+        Node<K, V> element = table[bucketIndex];
+        table[bucketIndex] = newNode(hash, key, value, element);
+        size++;
+    }
```

Comment:
```suggestion
    private void addNodeToBucket(K key, V value, Node<K,V> node) {
        node.next = new Node(key, value, null);
        size++;
    }
```

Reply:
but I think this method is redundant and better to move this logic inside put method